compressors/tANS.py

```diff
@@ -0,0 +1,380 @@
+        self.params = tans_params
+
+        ## build lookup tables
+        self.build_rans_base_decode_table()
+        self.build_expand_state_num_bits_table()
+
+    def build_rans_base_decode_table(self):
+        """
+        cache the rans_decoder.rans_base_decode_step function
+        """
+        rans_decoder = rANSDecoder(self.params)
+        self.base_decode_step_table = {}  # stores s, state_shrunk
+        for state in range(self.params.L, self.params.H + 1):
+            self.base_decode_step_table[state] = rans_decoder.rans_base_decode_step(state)
+
+    def build_expand_state_num_bits_table(self):
+        """
+        cache the expand state function
+        """
+        self.expand_state_num_bits_table = {}
+        for s in self.params.freqs.alphabet:
+            _min, _max = self.params.min_shrunk_state[s], self.params.max_shrunk_state[s]
+            for x_shrunk in range(_min, _max + 1):
+                num_bits = self.params.NUM_STATE_BITS - get_bit_width(x_shrunk)
+                self.expand_state_num_bits_table[x_shrunk] = num_bits
+
+    def decode_symbol(self, state: int, encoded_bitarray: BitArray):
+        # base rANS decoding step
+        s, state_shrunk = self.base_decode_step_table[state]
+
+        # remap the state into the acceptable range
+        num_bits = self.expand_state_num_bits_table[state_shrunk]
+        state_remainder = 0
+        if num_bits:
+            state_remainder = bitarray_to_uint(encoded_bitarray[:num_bits])
+        state = (state_shrunk << num_bits) + state_remainder
+
+        return s, state, num_bits
+
+    def decode_block(self, encoded_bitarray: BitArray):
+        # get data block size
+        data_block_size_bitarray = encoded_bitarray[: self.params.DATA_BLOCK_SIZE_BITS]
+        input_data_block_size = bitarray_to_uint(data_block_size_bitarray)
+        num_bits_consumed = self.params.DATA_BLOCK_SIZE_BITS
+
+        # get the final state
+        state = bitarray_to_uint(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + self.params.NUM_STATE_BITS]
+        )
+        num_bits_consumed += self.params.NUM_STATE_BITS
+
+        # perform the decoding
+        decoded_data_list = []
+        for _ in range(input_data_block_size):
+            s, state, num_symbol_bits = self.decode_symbol(
+                state, encoded_bitarray[num_bits_consumed:]
+            )
+
+            # rANS decoder decodes symbols in the reverse direction,
+            # so we add newly decoded symbol at the beginning
+            decoded_data_list = [s] + decoded_data_list
+            num_bits_consumed += num_symbol_bits
+
+        # Finally, as a sanity check, ensure that the end state should be equal to the initial state
+        assert state == self.params.INITIAL_STATE
+
+        return DataBlock(decoded_data_list), num_bits_consumed
+
+
+######################################## TESTS ##########################################
+
+
+def test_check_encoded_bitarray():
```

Comment:
Can we also explicitly test the generated tables here? Right now only checking encoded bits.
similarly might be useful to have a decoded table test. Can have a separate test_generated_tables()

Reply:
ll look into it.. if convenient!

Reply:
Assigned to @tpulkit

Reply:
Actually this is done done now @tpulkit